source/opt/ccp_pass.cpp

```diff
@@ -120,58 +120,37 @@ SSAPropagator::PropStatus CCPPass::VisitAssignment(ir::Instruction* instr) {
   }
 
-  // Otherwise, see if the RHS of the assignment folds into a constant value.
-  std::vector<uint32_t> cst_val_ids;
-  bool missing_constants = false;
-  bool varying_values = false;
-  instr->ForEachInId([this, &cst_val_ids, &missing_constants,
-                      &varying_values](uint32_t* op_id) {
+  // See if the RHS of the assignment folds into a constant value.
+  auto map_func = [this](uint32_t id) {
+    auto it = values_.find(id);
+    if (it == values_.end() || it->second == kVaryingSSAId) {
+      return id;
+    }
+    return it->second;
+  };
+  ir::Instruction* folded_inst = opt::FoldInstruction(instr, map_func);
+  if (folded_inst != nullptr && folded_inst->IsConstant()) {
+    values_[instr->result_id()] = folded_inst->result_id();
+    return SSAPropagator::kInteresting;
+  }
+
+  // If not, see if there is a least one unknown operand to the instruction.  If
+  // so, we might be able to fold it later.
+  bool could_be_improved = false;
+  instr->ForEachInId([this, &could_be_improved](uint32_t* op_id) {
     auto it = values_.find(*op_id);
     if (it == values_.end()) {
-      missing_constants = true;
-      return;
-    } else if (IsVaryingValue(it->second)) {
-      varying_values = true;
+      could_be_improved = true;
       return;
     }
-    cst_val_ids.push_back(it->second);
   });
-
-  // If we did not find a constant value for every operand in the instruction,
-  // do not bother folding it.  Indicate that this instruction does not produce
-  // an interesting value for now.
-  if (missing_constants) {
+  if (could_be_improved) {
     return SSAPropagator::kNotInteresting;
   }
 
-  // If we found at least one varying value, the instruction will never fold
-  // into anything interesting.  Mark it varying.
-  if (varying_values) {
-    return MarkInstructionVarying(instr);
-  }
-
-  auto constants = const_mgr_->GetConstantsFromIds(cst_val_ids);
-  assert(constants.size() != 0 && "Found undeclared constants");
-
-  // If any of the constants are not supported by the folder, we will not be
-  // able to produce a constant out of this instruction.  Consider it varying
-  // in that case.
-  if (!std::all_of(constants.begin(), constants.end(),
-                   [](const analysis::Constant* cst) {
-                     return IsFoldableConstant(cst);
-                   })) {
-    return MarkInstructionVarying(instr);
-  }
-
-  // Otherwise, fold the instruction with all the operands to produce a new
-  // constant.
-  uint32_t result_val = FoldScalars(instr->opcode(), constants);
-  const analysis::Constant* result_const =
-      const_mgr_->GetConstant(const_mgr_->GetType(instr), {result_val});
-  ir::Instruction* const_decl =
-      const_mgr_->GetDefiningInstruction(result_const);
-  values_[instr->result_id()] = const_decl->result_id();
-  return SSAPropagator::kInteresting;
+  // Otherwise, we will never be able to fold this instruction, so mark it
+  // varying.
+  return MarkInstructionVarying(instr);
```

Comment:
This seems really aggressive.

Reply:
This is less aggressive than it was before.  The previous code would say the instruction was varying if there was a least 1 input operand that was varying.  

Now, we will mark it as varying if every operand has moved from the top of the lattice (not in values_), and we were unable to fold.  At this point, things can only get worse for folding.  A constant will becoming bottom (varying).  If we can't fold to a constant now, then we will not be able to fold later.  So we might as well go straight to the bottom (varying).